File: ext/minify-html.js
import fs from 'node:fs';
import path from 'node:path';
import { Buffer } from 'node:buffer';
// import minifyHtml from '@minify-html/node';
import { minify } from '@minify-html/wasm';

const folders = [
  './gaya-hidup', './jejak-sejarah', './lainnya',
  './olah-media', './opini-sosial', './sistem-terbuka', './warta-tekno'
];

let stats = { 
  success: 0, 
  skipped: 0, 
  failed: 0, 
  errorList: [], 
  totalSaved: 0,
  totalBefore: 0,
  totalAfter: 0
};

// Helper untuk format ukuran byte agar manusiawi
const formatBytes = (bytes) => {
  if (bytes === 0) return '0 B';
  const k = 1024;
  const sizes = ['B', 'KB', 'MB'];
  const i = Math.floor(Math.log(bytes) / Math.log(k));
  return parseFloat((bytes / Math.pow(k, i)).toFixed(2)) + ' ' + sizes[i];
};

async function minifyFiles(dir) {
  if (!fs.existsSync(dir)) return;
  const files = fs.readdirSync(dir);

  for (const file of files) {
    const filePath = path.join(dir, file);
    const stat = fs.statSync(filePath);

    if (stat.isDirectory()) {
      await minifyFiles(filePath);
      continue;
    }

    if (!file.endsWith('.html') || file === 'index.html') continue;

    try {
      let originalHTML = fs.readFileSync(filePath, 'utf8');
      if (!originalHTML.trim()) continue;

      if (originalHTML.includes('udah_dijepit_oleh_Fakhrul_Rijal')) {
        stats.skipped++;
        continue;
      }

      // Ukuran Sebelum
      const sizeBefore = Buffer.byteLength(originalHTML, 'utf8');

      // --- PERBAIKAN KOMENTAR JS ---
      originalHTML = originalHTML.replace(/<script[\s\S]*?<\/script>/gi, (match) => {
        return match.replace(/^[ \t]*\/\/(?!#).*/gm, ''); 
      });

      const d = new Date();
      const tgl = d.toISOString().slice(0, 10);
      const minifySignature = `<noscript>udah_dijepit_oleh_Fakhrul_Rijal_${tgl}</noscript>`;

      const input = Buffer.from(originalHTML);
const output = minify(input, {
  allow_noncompliant_unquoted_attribute_values: true,
  allow_optimal_entities: true,
  allow_removing_spaces_between_attributes: true,
  collapse_whitespaces: true,
  ensure_spec_compliant_unquoted_attribute_values: false,
  keep_comments: false,
  keep_html_and_head_opening_tags: false,
  keep_spaces_between_attributes: false,
  minify_css: true,
  minify_doctype: true,
  minify_js: true, 
  remove_bangs: true,
  remove_processing_instructions: true,
});

// WASM return Uint8Array → convert ke string
const minifiedHTML = Buffer.from(output).toString('utf8') + minifySignature;
      const sizeAfter = Buffer.byteLength(minifiedHTML, 'utf8');
      const saved = sizeBefore - sizeAfter;

      fs.writeFileSync(filePath, minifiedHTML, 'utf8');
      
      stats.success++;
      stats.totalBefore += sizeBefore;
      stats.totalAfter += sizeAfter;
      stats.totalSaved += saved;

      const savingPercent = ((saved / sizeBefore) * 100).toFixed(1);
      console.log(`✅ [${savingPercent}%] : ${filePath} (${formatBytes(sizeBefore)} ➡️  ${formatBytes(sizeAfter)})`);

} catch (err) {
  stats.failed++;
  stats.errorList.push({ 
    path: filePath, 
    error: err?.message || String(err) 
  });
  console.error(`❌ Gagal jepit: ${filePath}`);
}
  }
}

async function run() {
  console.log('🧼 Memulai Minify Ultra (Mode Paralel)...');
  console.log('📂 Lokasi: Balikpapan | Status: Turbo On 🚀');
  
  const startTime = Date.now();

  try {
    await Promise.all(folders.map(f => minifyFiles(f)));
    
    const duration = (Date.now() - startTime) / 1000;
    const totalSavingPercent = stats.totalBefore > 0 
      ? ((stats.totalSaved / stats.totalBefore) * 100).toFixed(2) 
      : 0;

    console.log('\n' + '='.repeat(60));
    console.log('📊 REKAP PROSES LAYAR KOSONG (PARALEL)');
    console.log('='.repeat(60));
    console.log(`⏱️  Waktu Tempuh      : ${duration.toFixed(2)} detik`);
    console.log(`✅ Berhasil Dijepit  : ${stats.success} file`);
    console.log(`⏭️  Sudah Dijepit     : ${stats.skipped} file`);
    console.log(`❌ Gagal Proses      : ${stats.failed} file`);
    console.log('-'.repeat(60));
    console.log(`📉 Total Sebelum     : ${formatBytes(stats.totalBefore)}`);
    console.log(`📉 Total Sesudah     : ${formatBytes(stats.totalAfter)}`);
    console.log(`🚀 Ruang Dihemat     : ${formatBytes(stats.totalSaved)} (${totalSavingPercent}%)`);
    
    if (stats.failed > 0) {
      console.log('\n⚠️  DETAIL ERROR:');
      stats.errorList.forEach((item, i) => console.log(`${i+1}. ${item.path} -> ${item.error}`));
    }
    console.log('='.repeat(60) + '\n');
  } catch (err) {
    console.error('💥 Terjadi kesalahan saat eksekusi paralel:', err);
  }
}

run().catch(err => {
  console.error('💥 Fatal Error:', err);
  process.exit(1);
});

File: alat/minify-html.rs
use minify_html::{minify, Cfg};
use regex::Regex;
use std::fs;
use std::path::{Path, PathBuf};
use std::time::Instant;
use chrono::Local;
use rayon::prelude::*; // Supaya prosesnya paralel beneran
use std::sync::atomic::{AtomicUsize, Ordering};
use std::sync::Mutex;

// ======================================================
// KONFIGURASI
// ======================================================
const FOLDERS: &[&str] = &[
    "./gaya-hidup", "./jejak-sejarah", "./lainnya",
    "./olah-media", "./opini-sosial", "./sistem-terbuka", "./warta-tekno"
];

const SIGNATURE_KEY: &str = "udah_dijepit_oleh_Fakhrul_Rijal";

struct Stats {
    success: AtomicUsize,
    skipped: AtomicUsize,
    failed: AtomicUsize,
    total_before: AtomicUsize,
    total_after: AtomicUsize,
}

// Helper format bytes
fn format_bytes(bytes: usize) -> String {
    if bytes == 0 { return "0 B".to_string(); }
    let k = 1024.0;
    let sizes = ["B", "KB", "MB"];
    let i = (bytes as f64).log(k).floor() as usize;
    format!("{:.2} {}", (bytes as f64) / k.powi(i as i32), sizes[i])
}

fn minify_file(path: PathBuf, stats: &Stats, errors: &Mutex<Vec<String>>) {
    let file_name = path.file_name().unwrap().to_string_lossy();
    if !file_name.ends_with(".html") || file_name == "index.html" {
        return;
    }

    match fs::read_to_string(&path) {
        Ok(mut html) => {
            if html.trim().is_empty() { return; }
            if html.contains(SIGNATURE_KEY) {
                stats.skipped.fetch_add(1, Ordering::SeqCst);
                return;
            }

            let size_before = html.len();

            // --- PERBAIKAN KOMENTAR JS (RegEx ala Rust) ---
            let re_script = Regex::new(r"(?gi)<script[\s\S]*?>(?P<content>[\s\S]*?)</script>").unwrap();
            let re_comment = Regex::new(r"(?m)^[ \t]*//(?!#).*").unwrap();

            // Kita proses script tags untuk hapus comment //
            html = re_script.replace_all(&html, |caps: &regex::Captures| {
                let content = &caps["content"];
                let cleaned_content = re_comment.replace_all(content, "");
                caps[0].replace(content, &cleaned_content)
            }).to_string();

            // Setup Minify Config (Sama dengan config Node kamu)
            let mut cfg = Cfg::new();
            cfg.do_not_minify_doctype = false;
            cfg.ensure_spec_compliant_unquoted_attribute_values = false;
            cfg.keep_comments = false;
            cfg.minify_css = true;
            cfg.minify_js = true;
            cfg.remove_processing_instructions = true;
            cfg.collapse_whitespaces = true;

            let minified = minify(html.as_bytes(), &cfg);

            // Signature
            let tgl = Local::now().format("%Y-%m-%d").to_string();
            let signature = format!("<noscript>{}_{}</noscript>", SIGNATURE_KEY, tgl);

            let mut final_html = minified;
            final_html.extend_from_slice(signature.as_bytes());

            let size_after = final_html.len();

            if let Err(e) = fs::write(&path, final_html) {
                stats.failed.fetch_add(1, Ordering::SeqCst);
                errors.lock().unwrap().push(format!("{:?} -> {}", path, e));
            } else {
                let saved = if size_before > size_after { size_before - size_after } else { 0 };
                let percent = (saved as f64 / size_before as f64) * 100.0;

                println!("✅ [{:.1}%] : {:?} ({} ➡️ {})", percent, path.file_name().unwrap(), format_bytes(size_before), format_bytes(size_after));

                stats.success.fetch_add(1, Ordering::SeqCst);
                stats.total_before.fetch_add(size_before, Ordering::SeqCst);
                stats.total_after.fetch_add(size_after, Ordering::SeqCst);
            }
        },
        Err(e) => {
            stats.failed.fetch_add(1, Ordering::SeqCst);
            errors.lock().unwrap().push(format!("{:?} -> {}", path, e));
        }
    }
}

// Fungsi rekursif untuk ambil semua path file
fn get_all_files(dir: &Path, files: &mut Vec<PathBuf>) {
    if dir.is_dir() {
        if let Ok(entries) = fs::read_dir(dir) {
            for entry in entries.flatten() {
                let path = entry.path();
                if path.is_dir() {
                    get_all_files(&path, files);
                } else {
                    files.push(path);
                }
            }
        }
    }
}

fn main() {
    println!("🧼 Memulai Minify Ultra (Rust Parallel Engine)...");
    println!("📂 Lokasi: Balikpapan | Status: Turbo On 🚀");

    let start_time = Instant::now();
    let stats = Stats {
        success: AtomicUsize::new(0),
        skipped: AtomicUsize::new(0),
        failed: AtomicUsize::new(0),
        total_before: AtomicUsize::new(0),
        total_after: AtomicUsize::new(0),
    };
    let error_list = Mutex::new(Vec::new());

    // 1. Kumpulkan semua file dulu
    let mut all_files = Vec::new();
    for folder in FOLDERS {
        get_all_files(Path::new(folder), &mut all_files);
    }

    // 2. Proses secara paralel menggunakan Rayon (Mirip Promise.all tapi thread-based)
    all_files.into_par_iter().for_each(|path| {
        minify_file(path, &stats, &error_list);
    });

    let duration = start_time.elapsed().as_secs_f64();
    let before = stats.total_before.load(Ordering::SeqCst);
    let after = stats.total_after.load(Ordering::SeqCst);
    let saved = if before > after { before - after } else { 0 };
    let percent = if before > 0 { (saved as f64 / before as f64) * 100.0 } else { 0.0 };

    println!("\n{}", "=".repeat(60));
    println!("📊 REKAP PROSES LAYAR KOSONG (RUST PARALLEL)");
    println!("{}", "=".repeat(60));
    println!("⏱️  Waktu Tempuh      : {:.2} detik", duration);
    println!("✅ Berhasil Dijepit  : {} file", stats.success.load(Ordering::SeqCst));
    println!("⏭️  Sudah Dijepit      : {} file", stats.skipped.load(Ordering::SeqCst));
    println!("❌ Gagal Proses      : {} file", stats.failed.load(Ordering::SeqCst));
    println!("{}", "-".repeat(60));
    println!("📉 Total Sebelum      : {}", format_bytes(before));
    println!("📉 Total Sesudah      : {}", format_bytes(after));
    println!("🚀 Ruang Dihemat      : {} ({:.2}%)", format_bytes(saved), percent);

    let errs = error_list.lock().unwrap();
    if !errs.is_empty() {
        println!("\n⚠️  DETAIL ERROR:");
        for (i, err) in errs.iter().enumerate() {
            println!("{}. {}", i + 1, err);
        }
    }
    println!("{}\n", "=".repeat(60));
}
